### ISSUE TITLE
Unexpected `\` esacpe in json

### ISSUE THREAD

**Issue by whtsky** [huginn/huginn#2696]
Step to reproduce:

* Enter agent edit page
* Click on `Toggle View`, enter raw json edit mode
* Edit some field so it contains `\`
![图片](https://user-images.githubusercontent.com/876694/75566995-b3b29280-5a8b-11ea-8a09-fa61364aa63d.png)
* Click on `Toggle View` (Enter visual edit mode), then Click on `Toggle View` again (enter raw json edit mode)

Expected: `message` is still `a\nb`
Actual: `message` becomes `a\\nb`

![图片](https://user-images.githubusercontent.com/876694/75567076-e3619a80-5a8b-11ea-8be3-94d835ccac49.png)

---

Or even worse:

* Create a new agent, enter raw json edit mode, edit, carefully not to click on "Toggle View"
* Save the agent
* Edit the agent, do nothing, save the agent again

Expected: `message` is still `a\nb`
Actual: `message` becomes `a\\nb`

![123](https://user-images.githubusercontent.com/876694/75567298-43f0d780-5a8c-11ea-9bde-f7f763a14b65.gif)

**Comment by whtsky** [comment]
The same behavior can be observed on https://andrewcantino.com/examples/jsoneditor/example.html

I'm not sure if it's a bug or feature, but if it's a feature, it's better to have an option to disable it.

**Comment by dsander** [comment]
Ohboi, escaping is always fun! 😄 

I am not sure if this really is a bug. I am testing it with a `TriggerAgent`:

```json
{
  "expected_receive_period_in_days": "2",
  "keep_event": "false",
  "rules": [
    {
      "type": "regex",
      "value": "\\d+\\n",
      "path": "message"
    }
  ],
  "message": "Looks like your pattern matched in '{{value}}'!"
}
```
Editing the configuration with the "GUI" editor works as expected, I put in  \d+\n` as my regular expression. 

Sending this Event to the Agent properly triggers:

```
{
  "message": "31213\\n"
}
```

**Comment by whtsky** [comment]
My use case is that I'm trying to create some twitter publish automation in huginn.

What I want is (text showed in Twitter)
```
title
content
```

In Twitter Publish Agent, If I set to
```json
{
  "expected_update_period_in_days": "30",
  "message": "{{title}}\n{{content}}"
}
```

It would work as I expected. But the escape "feature" often makes my config
```diff
{
  "expected_update_period_in_days": "30",
-  "message": "{{title}}\n{{content}}"
+  "message": "{{title}}\\n{{content}}"
}
```

So the output would become (text showed in Twitter)
```
title\ncontent
```

As you can see, the escape problem really bugs me.

---

> Editing the configuration with the "GUI" editor works as expected, I put in \d+\n` as my regular expression.

The current escape feature is helpful when dealing with regexp, but not ideal if you want to deal with control characters, which in my case, line feeds.(`\n`)

**Comment by dsander** [comment]
I see, to be honest I am surprised that `"{{title}}\n{{content}}"` works, as far as I know it should be invalid JSON. If you want to insert an actual line break character you can use the `{% line_break %}` Liquid tag.

**Comment by whtsky** [comment]
> as far as I know it should be invalid JSON

It's a valid JSON. You can't put the actual line feed character ( the character you input by pressing the `Enter` button ) inside JSON strings, instead, you have to put `\n`.

So, 
```json
{
"message": "{{title}}
{{content}}"
}
```

is invalid, and 
```json
{
"message": "{{title}}\n{{content}}"
}
```

is valid.

---

> If you want to insert an actual line break character you can use the `{% line_break %}` Liquid tag.

Thanks for the tip! It sounds like a variable workaround, but I think the current escape behavior still needs some polish.

**Comment by virtadpt** [comment]
@dsander If it's not a bug, you can see bug-dom from there.  It can break saving agents sometimes, more often it messes up output because you see lots of \\n's in the text you get.

**Comment by dsander** [comment]
@virtadpt Sorry, I don't follow.


@whtsky 

> ```json
> ```json
> "message": "{{title}}
> {{content}}"
> }
> ```
> 
> 
> is invalid, and


Exactly my point if we would "convert" `\n` (the string) into a line break before parsing the JSON in the backend it would be invalid JSON.

The JSON editor always returns string, so `\n` is never converted to a line break. I think it converts `\n` to `\\n` when switching from the GUI to the raw JSON because a string like `"{\"test\": \"one\ntwo\"}"` would be invalid JSON.

**Comment by whtsky** [comment]
I'm not proposing huginn to convert `\n`(`\x5c\x6e`) into the real line break (`\x0a`), but saying the current GUI always convert `\n`(`\x5c\x6e`) into `\\n`(`\x5c\x5c\x6e`), which is a bug IMO.

---

![图片](https://user-images.githubusercontent.com/876694/75643193-2d5ba380-5c79-11ea-85d2-3e857120bfda.png)

Click `Toggle View`

![图片](https://user-images.githubusercontent.com/876694/75643202-35b3de80-5c79-11ea-8e1e-787834d977f4.png)

Click `Toggle View` again, I WOULD EXPECT `message` still be `a\nb`, but instead:

![图片](https://user-images.githubusercontent.com/876694/75643211-43696400-5c79-11ea-8047-25dee1cc7627.png)

**Comment by dsander** [comment]
I still think it's doing what it is supposed to do, what I don't understand is why it does not complain about invalid JSON when you enter

```json
{
"message": "{{title}}\n{{content}}"
}
```

as the RAW json because the same string isn't valid when passed to `JSON.parse` in JavaScript:

```
JSON.parse("{\"message\": \"{{title}}\n{{content}}\"}")

VM513:1 Uncaught SyntaxError: Unexpected token 
 in JSON at position 22
    at JSON.parse (<anonymous>)
    at <anonymous>:1:6
```

So I am assuming it converts `\n` to `\\n` before parsing the JSON.

What is interesting is that JavaScript and Ruby parses JSON differently:

```js
JSON.parse("{\"message\": \"{{title}}\\n{{content}}\"}")
{message: "{{title}}↵{{content}}"}
```

```ruby
irb(main):003:0> JSON.parse("{\"message\": \"{{title}}\\n{{content}}\"}")
=> {"message"=>"{{title}}\n{{content}}"}
```

**Comment by whtsky** [comment]
I see.

`"{\"message\": \"{{title}}\n{{content}}\"}"` is
```
{"message": "{{title}}
{{content}}"}
```
, which is invalid json.

![图片](https://user-images.githubusercontent.com/876694/75753957-ccaa9480-5d66-11ea-88fa-da1b0b6a9243.png)
(I'm not familier with Ruby, sorry. Ruby should has the same output.)

---

The correct string represent should be `"{\"message\": \"{{title}}\\n{{content}}\"}"`
![图片](https://user-images.githubusercontent.com/876694/75754137-1f844c00-5d67-11ea-9d49-8cf8e54fe4fb.png)

---

![图片](https://user-images.githubusercontent.com/876694/75754760-3aa38b80-5d68-11ea-824e-eb04e0fc11f8.png)

**Comment by dsander** [comment]
> `"{\"message\": \"{{title}}\n{{content}}\"}"` is
> 
> ```
> {"message": "{{title}}
> {{content}}"}
> ```
> 
> , which is invalid json.

Yes, so I think the JSON editor works around that error by injecting the additional `\`. But the way I see it the outcome doesn't really change because when the JSON is parsed in ruby the result is `\n` (the string) again.

**Comment by whtsky** [comment]
```json
{
  "message": "a\nb"
}
```
is a VALID json.

```json
{
  "message": "a
b"
}
```
is a INVALID json.

---

When I create a agent with options
```json
{
"expected_update_period_in_days": "10",
"message": "a\nb"
}
```
![图片](https://user-images.githubusercontent.com/876694/75769953-fcb36100-5d81-11ea-8b32-fbd65c09948c.png)

Huginn stores it as
```
"{\n\"expected_update_period_in_days\": \"10\",\n\"message\": \"a\\nb\"\n}"
```

Which can be parsed into a valid JSON.

---

What the JSON editor doing now, is change
```json
{
  "message": "a\nb"
}
```

Into
```json
{
  "message": "a\\nb"
}
```

**Comment by whtsky** [comment]
You can try to create a RSS Agent and set the url to `https://www.douban.com/feed/people/whtsky/interests`

Full options:
```json
{
  "expected_update_period_in_days": "30",
  "clean": "false",
  "url": "https://www.douban.com/feed/people/whtsky/interests"
}
```

then run it. See the events it created.
![图片](https://user-images.githubusercontent.com/876694/75770721-7861dd80-5d83-11ea-876a-034a677b3085.png)

As you can see, there are a lot of `\n` (TEXT) inside the payload(which is a JSON). And it's a valid JSON. That's what I'm talk about.

**Comment by whtsky** [comment]
https://gist.github.com/whtsky/0e0a1c1f20850953356678ffe627eff1#file-a-really-valid-json You can try to download this file and parse it in Ruby. Ruby should parse it without any complains.

**Comment by dsander** [comment]
Sorry I totally ignored the fact the saving the JSON in the plain text field actually saves it as `\n` in the database, for some reason I was under the impression that it also gets converted before the data is submitted.

https://github.com/cantino/jsoneditor/commit/2be5d5bb9e146b4c21ddfea78602addfeb6e81c8 this seems to be the commit that introduced the escaping. My suspicion is that the escaping is done because of the way `JSON.parse` works in JavaScript:

```
{
  "message": "a\nb"
}
```
```
> JSON.parse($("#agent_options").val())
{message: "a↵b"}
```

If the contents of `message` would then be put into the GUI editor it would contain an actual line break (which isn't possible because the fields only support a single line).

At some point the contents of the fields are encoded again which results in the double `\`. 

In short, I am now convinced that you are right and it is a bug, but I don't know if we can fix it. If we would change the behavior editing an Agent would change the configuration without the user doing anything. With the bug fixed having `a\nb` in the GUI editor would actually be saved as `a\nb` and not as `a\\nb` how it's currently done.

**Comment by whtsky** [comment]
> If the contents of message would then be put into the GUI editor it would contain an actual line break (which isn't possible because the fields only support a single line).

I understand it's hard to fix this without doing break change to the current escape behavior.

The most feasible way I can think of is to use textarea instead: https://jsfiddle.net/Lr70fq94/1/ .

https://github.com/cantino/jsoneditor looks like a javascript project but the Gemfile part looks scary to me :(

**Comment by dsander** [comment]
I don't think we have to use a text area, we would need to change the [escape logic here](https://github.com/cantino/jsoneditor/blob/master/lib/js/jquery.json-editor.js.coffee#L253). The main issue I have with it is that I have no clue why it was added in the first place :/.

I think the project uses ruby to run the tests and compile coffescript to javascript.
